src/lpf2hub.ts

```diff
@@ -439,174 +453,182 @@ export class LPF2Hub extends Hub {
-                    const accelY = Math.round((data.readInt16LE(6) / 28571) * 2000);
-                    const accelZ = Math.round((data.readInt16LE(8) / 28571) * 2000);
-                    /**
-                     * Emits when accelerometer detects movement. Measured in DPS - degrees per second.
-                     * @event LPF2Hub#accel
-                     * @param {string} port
-                     * @param {number} x
-                     * @param {number} y
-                     * @param {number} z
-                     */
-                    this.emit("accel", "ACCEL", accelX, accelY, accelZ);
-                    break;
-                }
-                case Consts.DeviceType.BOOST_TILT: {
-                    const tiltX = data.readInt8(4);
-                    const tiltY = data.readInt8(5);
-                    this._lastTiltX = tiltX;
-                    this._lastTiltY = tiltY;
-                    this.emit("tilt", port.id, this._lastTiltX, this._lastTiltY, this._lastTiltZ);
-                    break;
-                }
-                case Consts.DeviceType.POWERED_UP_REMOTE_BUTTON: {
-                    switch (data[4]) {
-                        case 0x01: {
-                            this.emit("button", port.id, Consts.ButtonState.UP);
-                            break;
-                        }
-                        case 0xff: {
-                            this.emit("button", port.id, Consts.ButtonState.DOWN);
-                            break;
-                        }
-                        case 0x7f: {
-                            this.emit("button", port.id, Consts.ButtonState.STOP);
-                            break;
-                        }
-                        case 0x00: {
-                            this.emit("button", port.id, Consts.ButtonState.RELEASED);
-                            break;
-                        }
-                    }
-                    break;
-                }
-                case Consts.DeviceType.DUPLO_TRAIN_BASE_COLOR: {
-                    if (data[4] <= 10) {
-                        this.emit("color", port.id, data[4]);
-                    }
-                    break;
-                }
-                case Consts.DeviceType.DUPLO_TRAIN_BASE_SPEEDOMETER: {
-                    /**
-                     * Emits on a speed change.
-                     * @event LPF2Hub#speed
-                     * @param {string} port
-                     * @param {number} speed
-                     */
-                    const speed = data.readInt16LE(4);
-                    this.emit("speed", port.id, speed);
-                    break;
+        if (port && port.connected && port.mode && port.modes[port.mode]) {
+            const mode = port.modes[port.mode];
+            const values = [];
+
+            for (let index = 4; index < data.length; index += Consts.VALUE_SIZE[mode.valueType]) {
```

Comment:
Port mode informations refer to signed values, I think they could add more types in the future and we will have to follow. It makes me think that it should have some "unsupported value type" to prevent misreading.

Reply:
I couldn't find any official info about those values being signed or unsigned. [This](https://lego.github.io/lego-ble-wireless-protocol-docs/index.html#value-format) does not mention it and [this](https://lego.github.io/lego-ble-wireless-protocol-docs/index.html#port-value-single) specifically refers to unsigned, which is not true in many cases.
From the empirical point of view there is a number of 8 bit port modes with raw max 255 and 16 bit with max 65535. Mostly unused modes, most relevant of them being LED RGB values (0-255), but this is only output mode.

Reply:
I can't retreive from where I understood it was signed so I must be wrong. I think we could rely on the min value and if it is under 0 then it is signed.